File: backend/app/schemas/catalogue.py
import uuid
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class VariantSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    sku: str
    name: str
    price: Decimal
    available_quantity: int


class ProductSummary(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    slug: str
    category: str | None
    brand: str | None
    image_url: str | None
    base_price: Decimal


class ProductDetail(ProductSummary):
    description: str | None
    is_active: bool
    variants: list[VariantSummary]


class ProductPage(BaseModel):
    items: list[ProductSummary]
    total: int
    limit: int
    offset: int


class ProductWrite(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = None
    category: str | None = Field(default=None, max_length=80)
    brand: str | None = Field(default=None, max_length=120)
    image_url: str | None = Field(default=None, max_length=500)
    base_price: Decimal = Field(ge=0, decimal_places=2)
    is_active: bool = True


class ProductUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = None
    category: str | None = Field(default=None, max_length=80)
    brand: str | None = Field(default=None, max_length=120)
    image_url: str | None = Field(default=None, max_length=500)
    base_price: Decimal | None = Field(default=None, ge=0, decimal_places=2)
    is_active: bool | None = None


class StockWrite(BaseModel):
    variant_id: uuid.UUID
    total_quantity: int = Field(ge=0)


class StockLevel(BaseModel):
    variant_id: uuid.UUID
    sku: str
    total_quantity: int
    reserved_quantity: int
    sold_quantity: int
    available_quantity: int

File: frontend/src/lib/catalogue.ts
import { z } from "zod";

import { request } from "./api";

export const productSummarySchema = z.object({
  id: z.string(),
  name: z.string(),
  slug: z.string(),
  category: z.string().nullable(),
  brand: z.string().nullable(),
  image_url: z.string().nullable(),
  base_price: z.string(),
});

export const variantSchema = z.object({
  id: z.string(),
  sku: z.string(),
  name: z.string(),
  price: z.string(),
  available_quantity: z.number(),
});

export const productDetailSchema = productSummarySchema.extend({
  description: z.string().nullable(),
  is_active: z.boolean(),
  variants: z.array(variantSchema),
});

export const productPageSchema = z.object({
  items: z.array(productSummarySchema),
  total: z.number(),
  limit: z.number(),
  offset: z.number(),
});

export type ProductSummary = z.infer<typeof productSummarySchema>;
export type ProductDetail = z.infer<typeof productDetailSchema>;
export type ProductPage = z.infer<typeof productPageSchema>;

export type BrowseFilters = {
  search?: string;
  category?: string;
  limit?: number;
  offset?: number;
};

export function browseProducts(filters: BrowseFilters = {}): Promise<ProductPage> {
  const params = new URLSearchParams();
  if (filters.search) params.set("search", filters.search);
  if (filters.category) params.set("category", filters.category);
  params.set("limit", String(filters.limit ?? 24));
  params.set("offset", String(filters.offset ?? 0));

  return request(`/products?${params}`, productPageSchema);
}

export function fetchProduct(slug: string): Promise<ProductDetail> {
  return request(`/products/${encodeURIComponent(slug)}`, productDetailSchema);
}

export function fetchCategories(): Promise<string[]> {
  return request("/products/categories", z.array(z.string()));
}

/** Prices arrive as strings so nothing is rounded before it is shown, and the
 *  pence are kept because a shop that hides them looks careless. */
export function formatPrice(amount: string): string {
  const value = Number(amount);
  if (!Number.isFinite(value)) {
    return amount;
  }
  return value.toLocaleString("en-US", {
    style: "currency",
    currency: "USD",
    minimumFractionDigits: 2,
  });
}
